src/test/app/Cron_test.cpp

```diff
@@ -0,0 +1,278 @@
+    }
+
+    void
+    testInvalidPreclaim(FeatureBitset features)
+    {
+        testcase("invalid preclaim");
+        using namespace test::jtx;
+        using namespace std::literals;
+
+        // no preclaim checks exists
+        BEAST_EXPECT(true);
+    }
+
+    void
+    testDoApply(FeatureBitset features)
+    {
+        testcase("doApply");
+        using namespace jtx;
+        using namespace std::literals::chrono_literals;
+        auto const alice = Account("alice");
+        Env env{*this, features | featureCron};
+
+        env.fund(XRP(1000), alice);
+        env.close();
+
+        auto const aliceOwnerCount = ownerCount(env, alice);
+
+        // create cron
+        env(cron::set(alice),
+            cron::delay(356 * 24 * 60 * 60),
+            cron::repeat(256),
+            fee(XRP(1)),
+            ter(tesSUCCESS));
+        env.close();
+
+        // increment owner count
+        BEAST_EXPECT(ownerCount(env, alice) == aliceOwnerCount + 1);
+
+        auto const accSle = env.le(keylet::account(alice.id()));
+        BEAST_EXPECT(accSle);
+        BEAST_EXPECT(accSle->isFieldPresent(sfCron));
+
+        auto const cronKey = keylet::child(accSle->getFieldH256(sfCron));
+        auto const cronSle = env.le(cronKey);
+        BEAST_EXPECT(cronSle);
+        BEAST_EXPECT(
+            cronSle->getFieldU32(sfDelaySeconds) == 356 * 24 * 60 * 60);
+        BEAST_EXPECT(cronSle->getFieldU32(sfRepeatCount) == 256);
+
+        // update cron
+        env(cron::set(alice),
+            cron::delay(100),
+            cron::repeat(10),
+            fee(XRP(1)),
+            ter(tesSUCCESS));
+        env.close();
+
+        // owner count does not change
+        BEAST_EXPECT(ownerCount(env, alice) == aliceOwnerCount + 1);
+
+        auto const accSle2 = env.le(keylet::account(alice.id()));
+        BEAST_EXPECT(accSle2);
+        BEAST_EXPECT(accSle2->isFieldPresent(sfCron));
+
+        // old cron sle is deleted
+        BEAST_EXPECT(!env.le(cronKey));
+
+        auto const cronKey2 = keylet::child(accSle2->getFieldH256(sfCron));
+        auto const cronSle2 = env.le(cronKey2);
+        BEAST_EXPECT(cronSle2);
+        BEAST_EXPECT(cronSle2->getFieldU32(sfDelaySeconds) == 100);
+        BEAST_EXPECT(cronSle2->getFieldU32(sfRepeatCount) == 10);
+
+        // delete cron
+        env(cron::set(alice), fee(XRP(1)), ter(tesSUCCESS));
+        env.close();
+
+        // owner count decremented
+        BEAST_EXPECT(ownerCount(env, alice) == aliceOwnerCount);
+
+        auto const accSle3 = env.le(keylet::account(alice.id()));
+        BEAST_EXPECT(accSle3);
+        BEAST_EXPECT(!accSle3->isFieldPresent(sfCron));
+
+        // old cron sle is deleted
+        BEAST_EXPECT(!env.le(cronKey2));
+    }
+
+    void
+    testWithFeats(FeatureBitset features)
+    {
+        testEnabled(features);
+        testFee(features);
+        testInvalidPreflight(features);
+        testInvalidPreclaim(features);
+        testDoApply(features);
```

Comment:
Thanks for doing the tests but I think we need also to actually test the ttCRON insertion after the delay is satisfied